### ISSUE TITLE
Wrong diagnostic when using constant of TAIT type in pattern

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#102011]
The test src/test/ui/type-alias-impl-trait/structural-match-no-leak.rs (copied below to preserve its state)

```rust
#![feature(type_alias_impl_trait)]

type Bar = impl Send;

// While i32 is structural-match, we do not want to leak this information.
// (See https://github.com/rust-lang/rust/issues/72156)
const fn leak_free() -> Bar {
    7i32
}
const LEAK_FREE: Bar = leak_free();

fn leak_free_test() {
    match LEAK_FREE {
        LEAK_FREE => (),
        //~^ ERROR constant pattern depends on a generic parameter
        //~| ERROR constant pattern depends on a generic parameter
        _ => (),
    }
}

fn main() {}
```

reports an error about "depends on a generic parameter", when there are no generic parameters anywhere. The issue is that we only have a single `TooGeneric` variant in https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/enum.ErrorHandled.html#variant.TooGeneric . We should change that fieldless variant to `TooGeneric(TooGeneric)` and introduce an

```rust
enum TooGeneric {
    /// The query encountered a generic parameter and could
    /// not extract required information from it.
    Param,
    /// The query encountered an opaque type in `Reveal::Userfacing`
    /// mode and was thus unable to reveal the hidden type.
    OpaqueType,
}
```

_Originally posted by @lcnr in https://github.com/rust-lang/rust/pull/101478#discussion_r963570994_

**Comment by lcnr** [comment]
i do think we should experiment with adding the too generic type to `TooGeneric`, i tried splitting that variant which was more annoying than helpful

**Comment by inquisitivecrystal** [comment]
Triage note: I was going to add one or more of the `D-` family labels, but wasn't sure which ones were appropriate. The OP and linked post don't really help me to understand what's going on and why it's a problem. I'm sure they're perfectly comprehensible to those who understand the situation better than I do, but a little bit of elaboration might help make this issue clearer. :)

**Comment by oli-obk** [comment]
Oof, yea, sorry about that. I added some explanation including what should actually be done.

**Comment by lcnr** [comment]
my idea was 
```rust
pub enum ErrorHandled<'tcx> {
    Reported(ErrorGuaranteed),
    Linted,
    TooGeneric(Ty<'tcx>),
}
```
`Ty<'tcx>` might not be perfect, e.g. sometimes we fail to resolve an instance, so there using `Ty<'tcx>` is wrong :thinking: maybe `TooGeneric(&'tcx TooGenericCause<'tcx>)` with `TooGenericCause<'tcx>` being an enum with the variants we care about :thinking: 

this should give us pretty good error messages with a hopefully acceptable perf impact

**Comment by compiler-errors** [comment]
This no longer reports the same error, and instead just says:

```
error: `Bar` cannot be used in patterns
  --> src/main.rs:15:9
   |
15 |         LEAK_FREE => (),
   |         ^^^^^^^^^

warning: unreachable pattern
  --> src/main.rs:16:9
   |
15 |         LEAK_FREE => (),
   |         --------- matches any value
16 |         _ => (),
   |         ^ unreachable pattern
   |
   = note: `#[warn(unreachable_patterns)]` on by default

warning: `playground` (bin "playground") generated 1 warning
error: could not compile `playground` due to previous error; 1 warning emitted
```

**Comment by JohnTitor** [comment]
Triage: Since we already have a test as `src/test/ui/type-alias-impl-trait/structural-match-no-leak.rs`, I think we could just close this issue. Closing.
